src/rebar_git_resource.erl

```diff
@@ -107,35 +107,72 @@ download(Dir, {git, Url, ""}, State) ->
     download(Dir, {git, Url, {branch, "master"}}, State);
 download(Dir, {git, Url, {branch, Branch}}, _State) ->
     ok = filelib:ensure_dir(Dir),
-    rebar_utils:sh(?FMT("git clone ~s ~s -b ~s --single-branch",
-                       [rebar_utils:escape_chars(Url),
-                        rebar_utils:escape_chars(filename:basename(Dir)),
-                        rebar_utils:escape_chars(Branch)]),
-                   [{cd, filename:dirname(Dir)}]);
+    git_clone(branch, git_vsn(), Url, Dir, Branch);
 download(Dir, {git, Url, {tag, Tag}}, _State) ->
     ok = filelib:ensure_dir(Dir),
-    rebar_utils:sh(?FMT("git clone ~s ~s -b ~s --single-branch",
-                       [rebar_utils:escape_chars(Url),
-                        rebar_utils:escape_chars(filename:basename(Dir)),
-                        rebar_utils:escape_chars(Tag)]),
-                   [{cd, filename:dirname(Dir)}]);
+    git_clone(tag, git_vsn(), Url, Dir, Tag);
 download(Dir, {git, Url, {ref, Ref}}, _State) ->
     ok = filelib:ensure_dir(Dir),
+    git_clone(ref, git_vsn(), Url, Dir, Ref);
+download(Dir, {git, Url, Rev}, _State) ->
+    ?WARN("WARNING: It is recommended to use {branch, Name}, {tag, Tag} or {ref, Ref}, otherwise updating the dep may not work as expected.", []),
+    ok = filelib:ensure_dir(Dir),
+    git_clone(rev, git_vsn(), Url, Dir, Rev).
+
+%% Use different git clone commands depending on git --version
+git_clone(branch,Vsn,Url,Dir,Branch) when Vsn >= {1,8,0} ->
+    rebar_utils:sh(?FMT("git clone ~s ~s -b ~s --single-branch",
+                        [rebar_utils:escape_chars(Url),
+                         rebar_utils:escape_chars(filename:basename(Dir)),
+                         rebar_utils:escape_chars(Branch)]),
+                   [{cd, filename:dirname(Dir)}]);
+git_clone(branch,_Vsn,Url,Dir,Branch) ->
+    rebar_utils:sh(?FMT("git clone ~s ~s -b ~s",
+                        [rebar_utils:escape_chars(Url),
+                         rebar_utils:escape_chars(filename:basename(Dir)),
+                         rebar_utils:escape_chars(Branch)]),
+                   [{cd, filename:dirname(Dir)}]);
+git_clone(tag,Vsn,Url,Dir,Tag) when Vsn >= {1,8,0} ->
```

Comment:
I think we should reverse that check so that if we have an undefined version (i.e. someone compiling by source or a future version of git) we should default to the current behaviour; when versions are identifiable as older, then we drop --single-branch as an option.